### PR TITLE
Add `cwd` option

### LINKED ISSUE #45: FR: cwd option
I'd love to have a `cwd` option similar to `cpy`

### PR DESCRIPTION
**Pull request by jopemachine**
Fixes #45.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -49,11 +49,23 @@ const cpFileAsync = async (source, destination, options, progressEmitter) => {
 	}
 };
 
+const changeCurrentWorkingDirectory = (cwd) => {
+	try {
+		process.chdir(cwd);
```
Comment (sindresorhus) [inline index.js:54]: You should never call `process.chdir` in packages. It applies to everything running in the process.

The correct approach is to transform all the paths.

File: index.d.ts [hunk 2]
```diff
@@ -15,6 +15,13 @@ declare namespace cpFile {
 		@default 0o777
 		*/
 		readonly directoryMode?: number;
+
+		/**
+		Working directory to find source file.
```
Comment (sindresorhus) [inline index.d.ts:20]: It should apply to the destination parameter too.

File: test/async.js [hunk 3]
```diff
@@ -267,3 +267,12 @@ test.serial('rethrow read after open errors', async t => {
 
 	Object.assign(fs, {createWriteStream, createReadStream});
 });
+
+test.serial('test cwd option', async t => {
```
Comment (sindresorhus) [inline test/async.js:271]: Should not use `.serial`.

File: test/async.js [hunk 4]
```diff
@@ -267,3 +267,12 @@ test.serial('rethrow read after open errors', async t => {
...
+	const error = await t.throwsAsync(cpFile('sync.js', t.context.destination));
+
+	t.is(error.name, 'CpFileError');
+	t.is(error.code, 'ENOENT');
+
+	await t.notThrowsAsync(cpFile('sync.js', t.context.destination, {cwd: './test'}));
```
Comment (sindresorhus) [inline test/async.js:277]:
```suggestion
	await t.notThrowsAsync(cpFile('sync.js', t.context.destination, {cwd: 'test'}));
```

File: index.js [hunk 5]
```diff
@@ -49,12 +49,14 @@ const cpFileAsync = async (source, destination, options, progressEmitter) => {
...
-		throw new CpFileError(error);
-	}
+const resolvePath = (cwd, sourcePath, destinationPath) => {
+	sourcePath = path.resolve(cwd, sourcePath);
+	destinationPath = path.resolve(cwd, destinationPath);
+
+	return [
+		sourcePath,
+		destinationPath
+	];
```
Comment (sindresorhus) [inline index.js:59]: Would be better to use an object.

File: index.js [hunk 6]
```diff
@@ -49,11 +49,25 @@ const cpFileAsync = async (source, destination, options, progressEmitter) => {
...
+};
+
 const cpFile = (sourcePath, destinationPath, options) => {
```
Comment (sindresorhus) [inline index.js:62]:
```suggestion
const cpFile = (sourcePath, destinationPath, options = {}) => {
```

File: test/sync.js [hunk 7]
```diff
@@ -216,3 +216,16 @@ test('rethrow utimes errors', t => {
 
 	fs.utimesSync.restore();
 });
+
+test('test cwd option', t => {
```
Comment (sindresorhus) [inline test/sync.js:220]: The word "test" is moot

File: index.d.ts [hunk 8]
```diff
@@ -15,6 +15,15 @@ declare namespace cpFile {
 		@default 0o777
 		*/
 		readonly directoryMode?: number;
+
+		/**
+		Base directory to find the source file and copy it to this directory.
+
+		`cwd` works by being prepended with source, destination path.
```
Comment (sindresorhus) [inline index.d.ts:22]: That's an implementation detail. It doesn't matter for the user. You should explain how it works from the user's perspective.

**Comment by jopemachine** [comment]
> CI is failing

It seems there are different issues with this.
I will try to make separate PRs handling each issue.

**Comment by jopemachine** [comment]
> > CI is failing
> 
> It seems there are different issues with this. I will try to make separate PRs handling each issue.

I made #48, #50. I think CI issue could be resolved with these.

File: readme.md [hunk 9]
```diff
@@ -59,6 +59,15 @@ Default: `true`
...
 Overwrite existing destination file.
 
+##### cwd
+
+Type: `string`\
+Default: `process.cwd()`
+
+Base directory when you use relative path to the source, destination paths.
+
+`cp-file` finds the source file from this path, and copy it to this directory.
```
Comment (sindresorhus) [inline readme.md:69]: You forgot to update index.d.ts.

Comment (sindresorhus) [reply readme.md:69]: This sentence is not even correct. That's not how `cwd` works. The first sentence is also not good enough.

File: readme.md [hunk 10]
```diff
@@ -59,6 +59,19 @@ Default: `true`
...
+Default: `process.cwd()`
+
+When your source, destination path are relative, `cp-file` basically finds out the source, destination paths from the path you executed the process (`process.dir()`)
+
+You can change this base path to which path you want.
+
+This ensures `cp-file`'s behavior be same regardless of your path executed the process.
+
+When your source, destination path are not relative, this option is ignored.
+
```
Comment (sindresorhus) [inline readme.md:74]: The docs here are still not good enough. I would recommend starting over. Try to succinctly explain what it does. You shouldn't need more than a couple of short sentences.

File: index.js [hunk 11]
```diff
@@ -85,23 +99,27 @@ const checkSourceIsFile = (stat, source) => {
...
 };
 
-module.exports.sync = (source, destination, options) => {
-	if (!source || !destination) {
+module.exports.sync = (sourcePath, destinationPath, options = {}) => {
+	if (!sourcePath || !destinationPath) {
 		throw new CpFileError('`source` and `destination` required');
 	}
 
+	if (options && options.cwd) {
```
Comment (sindresorhus) [inline index.js:107]:
```suggestion
	if (options.cwd) {
```

File: index.d.ts [hunk 12]
```diff
@@ -15,6 +15,13 @@ declare namespace cpFile {
 		@default 0o777
 		*/
 		readonly directoryMode?: number;
+
+		/**
+		Working directory to find source files.
```
Comment (sindresorhus) [inline index.d.ts:20]:
```suggestion
		The working directory to find source files.
		
		The source and destination path are relative to this.
```

How about this? Just to make sure we also mention the destination path, which can also be relative to this.

**Comment by sindresorhus** [comment]
It was implied that this applied to the other occurrence of this too: https://github.com/sindresorhus/cp-file/pull/46#discussion_r852088489

**Comment by danielweck** [comment]
I believe this PR shipped in `cp-file` v10 caused a regression in `cpy` v10 which in turn breaks `cpy-cli` v5:

https://github.com/sindresorhus/cpy/issues/107#issuecomment-1624307180

(it's a subtle edge case with relative paths and implicit `process.cwd()`, but it boils down to resolving against `options.cwd` where the correct previous behaviour was to resolve against `process.cwd()` ... I think the fix must be made on the consumer side of `cp-file`, i.e. `cpy`)

**Pull request merged**